File: Java/Medium/Accepted/walking-robot-simulation_6.java
class Solution {
    public int robotSim(int[] commands, int[][] obstacles) {
        
        int x = 0;
        int y = 0;
        int direction = 0;
        int max = 0;
        Set<String> obstacleSet = new HashSet<>();
        for (int[] obstacle: obstacles) {
            obstacleSet.add(obstacle[0] + " " + obstacle[1]);
        }
        for (int command: commands) {
            if (command == -2) {
                direction = (direction + 3) % 4;
            } else if (command == -1) {
                direction = (direction + 1) % 4;
            } else {
                for (int i = 0; i < command; i++) {
                    int nextX = x;
                    int nextY = y;
                    if (direction == 0) {
                        nextY++;
                    } else if (direction == 1) {
                        nextX++;
                    } else if (direction == 2) {
                        nextY--;
                    } else if (direction == 3) {
                        nextX--;
                    }
                    if (!obstacleSet.contains(nextX + " " + nextY)) {
                        x = nextX;
                        y = nextY;
                        max = Math.max(max, x * x + y * y);
                    } else {
                        break;
                    }
                }
            }
        }
        return max;
    }
}

File: Python3/Medium/Accepted/walking-robot-simulation_6.py
class Solution:
    def robotSim(self, commands: List[int], obstacles: List[List[int]]) -> int:
        
        # directions: 0, 1, 2, 3 (N, E, S, W)
        # x, y
        # (0, 1), (1, 0), (0, -1), (-1, 0)
        directions = [(0, 1), (1, 0), (0, -1), (-1, 0)]
        x, y = 0, 0
        direction = 0
        obstacles = set(map(tuple, obstacles))
        ans = 0
        for command in commands:
            if command == -2:
                direction = (direction - 1) % 4
            elif command == -1:
                direction = (direction + 1) % 4
            else:
                for _ in range(command):
                    nx, ny = x + directions[direction][0], y + directions[direction][1]
                    if (nx, ny) in obstacles:
                        break
                    x, y = nx, ny
                    ans = max(ans, x*x + y*y)
        return ans